sklearn/utils/__init__.py

```diff
@@ -213,12 +213,8 @@ def safe_indexing(X, indices):
                           DataConversionWarning)
             return X.copy().iloc[indices]
     elif hasattr(X, "shape"):
-        if hasattr(X, 'take') and (hasattr(indices, 'dtype') and
-                                   indices.dtype.kind == 'i'):
-            # This is often substantially faster than X[indices]
-            return X.take(indices, axis=0)
-        else:
-            return X[indices]
+        # just use fancy indexing, which is faster than numpy take (PR 14075)
```

Comment:
Uhm, are you sure that this is always the case. I always thought that `take` was usually slightly faster.

Reply:
Using an up-to-date python 3.7.3 and numpy 1.16.4 on Windows x64 and also using python 3.6.5 and numpy 1.16.4 on Linux x64 I see an immense slowdown with `X.take(indices, axis=0)` instead of `X[indices]`. Maybe `take ` was faster in an old numpy version? 

For me this significant speed difference looks like a bug in numpy, but I do not have the time to check old python and numpy versions and run this benchmark.

Reply:
I know that take was shown to be faster previously, especially with `mode='clip`. Take may be slower with memmaps for some reason?

Reply:
I use some benchmarks using the awesome timeit tool to isolate this behavior. Just add an empty `__init__.py` inside the benchmarks directory for the import to work.

**really slow take on numpy.memmap**
```
python -m timeit -n 100 -s "from numpy import arange; from benchmarks.bench_mnist import load_data; X_train, X_test, y_train, y_test = load_data(); indices=arange(200, dtype=int)" "X_train.take(indices, axis=0)"
100 loops, best of 3: 170 msec per loop
```

**fancy indexing on the same data is 100x faster**
```
python -m timeit -n 100 -s "from numpy import arange; from benchmarks.bench_mnist import load_data; X_train, X_test, y_train, y_test = load_data(); indices=arange(200, dtype=int)" "X_train[indices]"
100 loops, best of 3: 121 usec per loop
```

The transformation from np.memmap into np.ndarray makes the algorithm a bit faster.

**Still really slow usage of take**
```
python -m timeit -n 100 -s "from numpy import arange; from sklearn.utils.validation import check_X_y; from benchmarks.bench_mnist import load_data; X_train, X_test, y_train, y_test = load_data(); indices=arange(200, dtype=int); X_train, y_train = check_X_y(X_train, y_train);" "X_train.take(indices, axis=0)"
100 loops, best of 3: 157 msec per loop
```

**The fancy indexing also got faster than before**
```
python -m timeit -n 100 -s "from numpy import arange; from sklearn.utils.validation import check_X_y; from benchmarks.bench_mnist import load_data; X_train, X_test, y_train, y_test = load_data(); indices=arange(200, dtype=int); X_train, y_train = check_X_y(X_train, y_train);" "X_train[indices]"
100 loops, best of 3: 89.5 usec per loop
```

I am not able to construct an example which shoes a speed improvement for take. Although, in this case the time difference is really small and there might be no benefit at all.

```
python -m timeit -n 100 -s "from numpy import arange; X_train = arange(60000 * 100).reshape((60000, 100)); indices=arange(200, dtype=int);" "X_train.take(indices, axis=0)"
100 loops, best of 3: 25.8 usec per loop
```

```
python -m timeit -n 100 -s "from numpy import arange; X_train = arange(60000 * 100).reshape((60000, 100)); indices=arange(200, dtype=int);" "X_train[indices]"
100 loops, best of 3: 24.2 usec per loop
```

Really looks like a serious regression in numpy, which is exposed here.

Reply:
This is an important statement from @seberg given under the numpy issue.

> In most cases (I can give the details, but...) indexing outperforms take nowadays, due to optimizing the memory order of the operation.

Considering my benchmark results it does not make sense to define a case for take() and instead use fancy indexing all the time, at least in my opinion.

Reply:
Well, take has completely different code paths, and yes "clip" will speed it up a bit. There is an old myth that advanced indexing is just slow, which originated in some blog post by Wes or so and is outdated for years now.
That does not mean that take cannot be quite a bit faster. The most extreme example of this is likely: `arr = np.arange(10000).reshape(-1, 2); ind = np.random.randint(len(arr), size=2000); arr.take()`. The important part here is the `2`-shape, anything except two should have a smaller difference probably